[pre-push] Check publish pushes for references to redacted commits

JonWBedard opened the pull request:
#### f4a0a73081dd191d2c52963e85b85b0c99b29dbd
<pre>
[pre-push] Check publish pushes for references to redacted commits
<a href="https://bugs.webkit.org/show_bug.cgi?id=256025">https://bugs.webkit.org/show_bug.cgi?id=256025</a>
rdar://108588596

Reviewed by Elliott Williams.

When operating in &quot;publish&quot; mode, our pre-push hook should check commits
for redacted commits, even if the commit in question originates from a
different remote.

* Tools/Scripts/hooks/pre-push:

Canonical link: <a href="https://commits.webkit.org/263677@main">https://commits.webkit.org/263677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db789df3147b996b0dc6b603b727978f4dca60f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5420 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5420 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5788 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/5538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/6358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5522 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/5587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/4854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6991 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/3065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/4859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/12015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/4931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/6645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/5375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/5538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/4835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/8936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/614 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -298,8 +304,17 @@ def main(name, remote):
+            rs = remotes_with.get(commit, [])
+            if not rs:
+                continue
+            if name in rs or (MODE != PUBLISH_MODE and rs):
```
nit: `rs` is already truthy:
```suggestion
            if name in rs or MODE != PUBLISH_MODE:
```

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -204,30 +204,23 @@ def message_for(commit):
+        class_1_level = security_level_for_remotes(remotes, name_to_remote)
+    # The publication workflow allows class 1 violations, but we should still check for class 2 and 3 violations
+    if class_1_level == 0 or (not allow_class_1 and remotes):
+        return class_1_level, CLASS_1
```
nit: I believe this is logically equivalent, and is for me to follow.

```suggestion
        class_1_level = security_level_for_remotes(remotes, name_to_remote)
        # The publication workflow allows class 1 violations, but we should still check for class 2 and 3 violations
        if allow_class_1 and class_1_level > 0:
            pass
        else:
            return class_1_level, CLASS_1
```

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -204,30 +204,23 @@ def message_for(commit):
     message = message_for(commit)
     title = message.splitlines()[0]
 
-    # Compute class 3 before class 2, because a class 3 violation is more interesting than a class 2
-    # if both indicate a commit is potentially sensative
```
Thanks for moving this down.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -237,22 +230,35 @@ def security_level_for_commit(commit, remotes, name_to_remote=None):
             secondary = UNPACK_SECONDARY_RE.match(secondary).groups()[0]
         for candidate in [primary, secondary]:
-            original_remotes = remotes_for(candidate)
+            original_remotes = remotes_for(candidate.rstrip('.')) if candidate else []
```
`candidate.rstrip('.')`: What is this doing? Is it possible to move the `.` elision into the regular expression this variable comes from?

nit: I think this is more readable:
```python3
        for candidate in [primary, secondary]:
            if not candidate: 
                continue
            original_remotes = remotes_for(candidate.rstrip('.'))
```

or even:
```python3
        for candidate in filter(bool, (primary, secondary)):
            original_remotes = remotes_for(candidate.rstrip('.'))
```

emw-apple reviewed (approved):
r=me, though I feel like this code is getting challenging to follow, so I left some suggestions on improving control flow readability.

JonWBedard committed:
[pre-push] Check publish pushes for references to redacted commits
https://bugs.webkit.org/show_bug.cgi?id=256025
rdar://108588596

Reviewed by Elliott Williams.

When operating in "publish" mode, our pre-push hook should check commits
for redacted commits, even if the commit in question originates from a
different remote.

* Tools/Scripts/hooks/pre-push:

Canonical link: https://commits.webkit.org/263677@main
